### DISCUSSION TITLE
Thanks / java project can't find EnumEntriesKt class at runtime

### DISCUSSION THREAD

**Discussion by ninecatsj** [andregasser/bigbone#254] [General]
First - thank you for picking up this project, I have a small personal project that was based on mastodon4j, so it's really great to find a successor to it that's being actively developed and maintained.

I migrated that personal project (a simple mastodon bot, written in java, built using maven) from mastodon4j to the latest snapshot of BigBone this weekend, and it's all working well except for one build issue I found. It compiles without problems, but fails with a runtime exception because it can't find the kotlin.enums.EnumEntriesKt class:

```
Exception in thread "main" java.lang.NoClassDefFoundError: kotlin/enums/EnumEntriesKt
	at social.bigbone.api.entity.Status$Visibility.<clinit>(Status.kt:205)
	at org.ninecats.aedatebot.AncientEgyptianDateBot.main(AncientEgyptianDateBot.java:101)
Caused by: java.lang.ClassNotFoundException: kotlin.enums.EnumEntriesKt
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	... 2 more
```

After a little investigation, I noticed that the build was pulling in an older version of kotlin-stdlib - 1.6.20. If I add the following to my maven pom.xml, it avoids the runtime error:

```
<properties>
  <kotlin.version>1.9.10</kotlin.version>
</properties>

<dependencies>
  <dependency>
    <groupId>org.jetbrains.kotlin</groupId>
    <artifactId>kotlin-stdlib</artifactId>
    <version>${kotlin.version}</version>
  </dependency>
</dependencies>
```

I didn't see anything like that mentioned in the readme / sample code, so I wondered if you had any advice - did I miss something in my project, and is this the right way to fix it? Or is it something that should be addressed somehow in BigBone?

**Comment by andregasser** [comment]
Hello @ninecatsj !

First off, glad to hear that you enjoy using this library. And also thanks for bringing this issue up. I'll have a look and get back to you as soon as I have some more info available. I need to check a few things here and there.

Best regards,
André

**Comment by andregasser** [comment]
Hello @ninecatsj I think I have a fix for this, but I would like to make sure it really works. Could you eventually provide me a small snipped of your code that shows the problematic area in your code where the exception is thrown? Thank you very much.

**Reply by andregasser** [reply]
Update: I was able to reproduce it easily by creating a new Maven project that uses the latest BigBone snapshot as dependency. As code, I use the "PostStatusWithMediaAttached.java" sample from the BigBone project. When running this code, it terminates with the same error as you have reported. Will now test if my fix works. I keep you posted on the progress.

**Reply by ninecatsj** [reply]
Hi André! Thanks for looking into this issue.

My code was very similar to what you have in that sample java file, as I was using it as a guide when migrating from mastodon4j.

Hopefully your fix will work, but if you need any more info from me, please just let me know. I'm happy to share any snippets of code, or the pom.xml file, if that would help.

**Comment by andregasser** [comment]
Hello @ninecatsj,

I have pushed a new commit which fixed the issue for my reproducer project. https://github.com/andregasser/bigbone/commit/7cd449f4e2696c801ecbccbb23ff078a03517c1e

The problem was, that a wrong version for kotlin-stdlib and kotlin-stdlib-common was pulled in as a dependency (I think it was 1.6.20). EnumEntries.kt is a file which was newly introduced in stdlib version 1.8 and I think this is where the trouble started. So I had to make sure, that I use the same version >=1.8 for all the stdlib dependencies which are used by BigBone.

I was able to achieve this by specifying the version explicitly in the build.gradle files. You can see this in the commit above.

I will soon publish a new snapshot. Once this is out, I'd be happy to get some feedback from you.

![image](https://github.com/andregasser/bigbone/assets/535130/1492aa1f-9dc0-4c9d-9743-a17e94cc4670)

Best regards,
André

**Reply by andregasser** [reply]
New snapshot is out: https://github.com/andregasser/bigbone/releases/tag/v2.0.0-20230902.213631

**Reply by ninecatsj** [reply]
Hi @andregasser, I built my project against the latest snapshot this evening, and I can confirm I no longer need the workarounds in my pom.xml file - it's picking up a consistent set of dependencies, and there is no runtime error any more. Thanks very much for fixing this!

**Reply by andregasser** [reply]
Thanks for the positive feedback! I am glad your issue is solved 🙂

Should you run into other issues in the future, feel free to contact us again. Always happy to get constructive feedback of any kind.

Have a great time!
